docs: OEP-67 - Document our best practice tools and tech. (#518)
openedx/open-edx-proposals: oeps/best-practices/oep-0067-bp-tools-and-technology.rst

OEP-67: Tools and Technology Standards
######################################

.. list-table::
   :widths: 25 75

   * - OEP
     - Link to the doc in the following format::

        :doc:`OEP-67 <oep-0067-bp-tools-and-technology>`

   * - Title
     - Tools and Technology Standards
   * - Last Modified
     - 2023-09-05
   * - Authors
     - Feanil Patel <feanil@axim.org>
   * - Arbiter
     - TBD
   * - Status
     - Accepted
   * - Type
     - Best Practice
   * - Created
     - 2023-09-05
   * - Review Period
     - 2023-09-05 - 2023-10-xx
   * - Resolution
     - <links to any discussions where the final status was decided>
   * - References
     - :doc:`oep-0011-bp-FED-technology`

Abstract
********

The Open edX Platform is built using many 1st and 3rd party tools and technologies.
This document covers the current recommendations for the tools and technologies
that should be used.

Motivation
**********

> The Open edX platform has a lot of 1st and 3rd party tools and technology that it uses.

Reviewer: This ends up being pretty repetitive of the Abstract, may not be easily avoidable.

The problems we're trying to solve are:

    * It is difficult to know what the current best practice for a given area.

    * It is unclear how to go about the discussion and decision making process
      to change a best practice.

:doc:`OEP-11 <oep-0011-bp-FED-technology>` serves as a great example for the
usefulness of this kind of OEP but limits itself to be relevant only to
frontend technologies.  This OEP would supersede :doc:`OEP-11
<oep-0011-bp-FED-technology>` and encompass not only frontend technolog but any
technology that:

    * Is considered a best practice for the Open edX Platform

    * Is relevant to a large subset of the Open edX Platform

Specification
*************

.. note:: We'll fill this out before the OEP lands but leaving mostly blank for initial feedback.

General Technology Selection
============================

This technology is not specific to frontend or backend code.

* Codecov is an example in this category.

* Github actions is probably another

Frontend Technology Selection
=============================

> The phrase “frontend” is used to mean any part of the platform that is shown to

Reviewer: Should the definition just be a note?

users. This encompasses views rendered in Python on the server, interactive
interfaces written using JavaScript, and CSS styling.

This will be an exact copy of the content of the **Technology Selection**
section of OEP-11

Backend Technology Selection
============================

This section will cover any technology or tools that are not considered a part
of the frontend.

This section will be empty to begin with and I'll create a separate PR to add any
technologies that we consider standard today. This will probably include:

    * Django

    * Django Rest Framework

    * Python?

    * drf-yasg (A thing that is currently standard but that I hope to change)

    * tox

    * pytest

    * edx-lint/pylint

Decisions
*********

This section lists the decision records for any time a standard technology has
been change in the system along with the reasons why and the alternatives
considered.

> .. We'll have a toctree to an ADR folder with decisions.  We'll move all the existing fronted decisions to under this OEP.

Reviewer 2: Is the idea to have `oeps/best-practices/oep-0067/decisions/thing-we-decided.rst` or `oeps/best-practices/oep-0067/decisions/frontend/thing-we-decided.rst`?

Author: I had assumed the first one but I'm very open to the second option.  What's the value of the second option in your mind, I see that it provides a bit more separation but I'm not clear on what that separation buys us.

Reviewer 2: I'm fine either way. I guess I was thinking the separation would buy us a nested TOC. This was asked out of curiosity not opinion.

> Consequences

Reviewer: Add tech-radar consequences.

Author: In my mind, the radar might still be useful as a graphical interface for this data.  This might imply that we should add the radar states to the entries for each item in this oep so it can also convey provisional or hold items along with what is currently implicitly adopt items.

I'm curious to hear what others think about this.

usefulness of this kind of OEP but limits itself to be relevant only to
frontend technologies.  This OEP would supersede :doc:`OEP-11
</archived/oep-0011-bp-FED-technology>` and encompass not only front-end technology but any
technology that:

* Is considered a best practice for the Open edX Platform

* Is relevant to a large subset of the Open edX Platform

Specification
*************

.. note:: We'll fill this out before the OEP lands but leaving mostly blank for initial feedback.

General Technology Selection
============================

This technology is not specific to frontend or backend code.

#. **Codecov should be used to report code coverage by unit tests**

   **Rationale**: It is important to measure the amount of code covered by our
   automated test suites. By striving for a high level of test coverage, we can
   reduce the number of bugs that can only be found via manual testing, 
   and by using codecov to run coverage in CI,
   contributors are automatically reminded to include tests for any new code.

   The Open edX community has standardized around `Codecov`_ as a coverage reporting service.

#. **Use Github actions for CI Testing**

>    **Rationale**: As a healthy open source project we need good continouse integration

Reviewer 3: ```suggestion
   **Rationale**: As a healthy open source project we need good continuous integration
```

   testing to prevent the introduction of obvious bugs into the platform code.  The Github Actions

>    tool should be used to perform Continuous Integration(CI) testing for all Open edX repositories

Reviewer 3: ```suggestion
   tool should be used to perform Continuous Integration (CI) testing for all Open edX repositories
```

> OEP-67: Tools and Technology Standards
> ######################################

Reviewer 4: ```suggestion
OEP-67: Standard Tools and Technologies
#######################################
```

Echoing my other comment here, but consider a title that clarifies that this OEP is the place to pick which tools and technologies we use, not the place to discuss things like coding standards, a11y standard practices, etc.

   * - Review Period
     - 2023-09-05 - 2023-10-xx
   * - Resolution
     - <links to any discussions where the final status was decided>
   * - References
     - :doc:`oep-0011-bp-FED-technology`

Abstract
********

The Open edX Platform is built using many 1st and 3rd party tools and technologies.
This document covers the current recommendations for the tools and technologies
that should be used.

Motivation
**********

The Open edX platform has a lot of 1st and 3rd party tools and technology that it uses.
The problems we're trying to solve are:

    * It is difficult to know what the current best practice for a given area.

    * It is unclear how to go about the discussion and decision making process
      to change a best practice.

:doc:`OEP-11 <oep-0011-bp-FED-technology>` serves as a great example for the
usefulness of this kind of OEP but limits itself to be relevant only to
frontend technologies.  This OEP would supersede :doc:`OEP-11
<oep-0011-bp-FED-technology>` and encompass not only frontend technolog but any
technology that:

    * Is considered a best practice for the Open edX Platform

    * Is relevant to a large subset of the Open edX Platform

Specification

> *************

Reviewer 4: I like that you kept each of the Rationale sections within the Specification fairly brief. The linked ADRs will contain plenty of detail, I imagine, so to reduce duplication and ease readability, it will be good to just stick to summaries here.

> .. note:: We'll fill this out before the OEP lands but leaving mostly blank for initial feedback.

Reviewer 4: Consider ordering the spec so that the import stuff is generally closer to the top. Seeing Codecov as the first item in the spec feels a little off, given how minor of a choice it is in the grand scheme of things. Maybe having the subsection order be [Python, JS, General] would be more conducive to that.

Author: Thinking more about this, I think it would be fine to Move general to the bottom so the order would be [JS, Python, General]  let me know if others have opinions on this and I can make a decisions at some point based on feedback.

Reviewer 4: @feanil I do like that idea, but it's not blocking for me.

General Technology Selection
============================

This technology is not specific to frontend or backend code.

> #. **Codecov should be used to measure code covered by tests**
> 
>    **Rationale**: It is important to measure the amount of code covered by our
>    automated test suites. By striving for a high level of test coverage, we can
>    reduce the number of bugs that can only be found via manual testing. another
>    added benefit comes when we include coverage runs in the CI process, as
>    contributors are automatically reminded to include tests for any new code.
> 
>    The Open edX community has standardized around `Codecov`_ as a measurement tool.

Reviewer 4: ```suggestion
#. **Codecov should be used to report code coverage by unit tests**

   **Rationale**: It is important to measure the amount of code covered by our
   automated test suites. By striving for a high level of test coverage, we can
   reduce the number of bugs that can only be found via manual testing, 
   and by using codecov to run coverage in CI,
   contributors are automatically reminded to include tests for any new code.

   The Open edX community has standardized around `Codecov`_ as a coverage reporting service.
```

(Nitpicking this particular entry a bit, but I think this will be relevant for other technology choices too.)

Codecov feels different than, say, React, because it's a proprietary SaaS thing. If the offering starts to get bad, we will need to switch, even if we like the tools that come with the service. I think it's fine to list it alongside React et al, but I want to be clear that this isn't really a tool, it's a service.

And if Codecov does ever get bad, we can continue locally measuring Python code coverage using coverage.py (which might be what Codecov is using under the hood... not sure).

Author: I think it's useful to enumerate the specific services here because often many services can do the same job but picking one and listing it is a useful way to reduce fracturing across the ecosystem.  Also, if tomorrow Codecov got bad, we don't currently have a good place to have the conversation about what the alternatives are and why we would pick one over others.  I think this OEP will provide the place for that conversation for both tools and SaaS services we use as a community.

   tool should be used to perform Continuous Integration(CI) testing for all Open edX repositories
   that perform any sort of software testing.

Frontend Technology Selection
=============================

.. note:: 

   The phrase “frontend” is used to mean any part of the platform that is shown to
   users. This encompasses views rendered in Python on the server, interactive
   interfaces written using JavaScript, and CSS styling.
   
#. **Use React**

   **Rationale**: React must be used for building new UIs, as it is
   widely adopted by the community and strikes a balance between
   flexibility and feature richness.

   When building in existing Django server-rendered pages, one can use the
   react_render helper method to bridge to React. This bridge provides an
   easy way to pass data into a React component from Django via props. For
   more information about react_render see `ReactRender`_.

#. **Use Redux**

   For state management of complex
   client-side interactions, Redux must be used. This library was chosen
   because it sees strong use in the React community, but is also flexible
   enough to be used in situations where a hybrid React/Backbone architecture

>    exists.

Reviewer 4: I remember David Joy saying at one point that Redux wasn't worth the hassle and that MFE devs should feel free to simply not use it.

I'm not sure if I remember that right or whether that statement still applies, I'd need @arbrandes or @brian-smith-tcril 's input.

(I guess this is the sort of thing that will make this OEP useful :)

Author: This is a copy from the current frontend OEP.  I think any corrections should be made as follow-up ADRs and updates to the OEP.

   **Note**: edX previously used CoffeeScript, but its use has now been
   deprecated. Community interest in TypeScript has also grown, but it and
   other languages that do not follow the ECMA-262 spec should not be used.

#. **JavaScript code should follow the edX ESLint configuration**

   **Rationale**: In order to standardize and enforce Open edX's JavaScript
   coding style across multiple codebases, edX has published an ESLint
   configuration that provides an enforceable specification. The
   `edX ESLint Config`_ is made available as an npm package that can be
   installed into any Open edX package.

   **Exception**: The `edX ESLint Config for ES5`_ may be used only where ES5
   is in use. Both configs may be used in the same codebase through the
   use of `ESLint configuration cascading`_.

#. **JavaScript libraries should be installed via npm**

   **Rationale**: It is important that JavaScript libraries are kept
   up-to-date, and one key component is to make it as simple as possible
   to perform upgrades. Projects that use npm maintain all of their
   library requirements in a single `package.json` file, which can be
   easily changed as the versions change.

> #. **JavaScript projects should publish a package.lock file**

Reviewer 4: ```suggestion
#. **JavaScript projects should publish a package-lock file**
```

#. **JavaScript should be bundled using Webpack**

   **Rationale**: `Webpack`_ is the tool of choice for code optimization and
   bundling, as it is widely used throughout the industry, and because
   it seamlessly handles modern code bases as well as all of the older
   technologies used by edX, such as `AMD Modules`_. Webpack should be
   implemented to handle as much of the "asset pipeline" as possible,
   rather than passing this responsibility on to Django.

#. **JavaScript dependencies should be managed with ES2015 Modules**

   **Rationale**: JavaScript module systems allow front end code to specify
   its dependencies and be grouped into bundles that minimize the assets
   needed to provide page functionality. The most prevalent module syntax
   is currently `ES2015 Modules`_, which should be adopted everywhere
   edX code is written to the ES2015 spec or later.

   **Exception**: Much of edX's existing (ES5) JavaScript uses the older
   `AMD Modules`_ syntax for modules. AMD Modules are interoperable
   with ES2015 Modules if Webpack is used for bundling, so AMD is an
   acceptable module definition if the code must remain ES5.

> #. **CSS should be generated using Sass**

Reviewer 4: ```suggestion
#. **CSS should be generated using SCSS**
```

We want people to use the SCSS syntax and *not* the original Sass syntax, right? If so , let's put that on the first line.

Backend Technology Selection
============================

This section will cover any technology or tools that are not considered a part
of the frontend.

.. This section will be empty to begin with and I'll create a separate PR to add any
   technologies that we consider standard today. This will probably include:

    * Django

    * Django Rest Framework

    * Python?

    * drf-yasg (A thing that is currently standard but that I hope to change)

    * tox

    * pytest

    * edx-lint/pylint

Decisions
*********

This section lists the decision records for any time a standard technology has
been change in the system along with the reasons why and the alternatives
considered.

.. We'll have a toctree to an ADR folder with decisions.  We'll move all the existing fronted decisions to under this OEP.

.. toctree::
   :caption: Overall Decisions
   :maxdepth: 1
   :glob:

   oep-0067/decisions/overall/*

.. toctree::
   :caption: Frontend Decisions
   :maxdepth: 1
   :glob:

   oep-0067/decisions/frontend/*

.. toctree::
   :caption: Backend Decisions
   :maxdepth: 1
   :glob:

   oep-0067/decisions/backend/*

Consequences
************

* OEP-11 will be superseded.

> * All ADRs under OEP-11 will be moved to be under this OEP instead.

Reviewer 4: Yay for consolidation!

OEP-67: Standard Tools and Technologies
#######################################

.. list-table::
   :widths: 25 75

   * - OEP

>      - Link to the doc in the following format::
> 
>         :doc:`OEP-67 <oep-0067-bp-tools-and-technology>`

Reviewer 4: ```suggestion
     - :doc:`OEP-67 <oep-0067-bp-tools-and-technology>`
```

   * - Title
     - Tools and Technology Standards
   * - Last Modified
     - 2023-09-05
   * - Authors
     - Feanil Patel <feanil@axim.org>
   * - Arbiter
     - Braden MacDonald  <braden@opencraft.com>
   * - Status
     - Accepted
   * - Type
     - Best Practice
   * - Created
     - 2023-09-05
   * - Review Period
     - 2023-09-05 - 2023-10-10
   * - Resolution
     - * https://github.com/openedx/open-edx-proposals/pull/518
   * - References
     - :doc:`oep-0011-bp-FED-technology`

Abstract
********

> The Open edX Platform is built using many 1st and 3rd party tools and technologies.

Reviewer 5: As long as we are nitting:
```suggestion
The Open edX Platform is built using many first- and third-party tools and technologies.
```

This document covers the current recommendations for which tools and technologies should be used.

Motivation
**********

The Open edX platform has a lot of first and third party tools and technology that it uses.
The problems we're trying to solve are:

>     * It is difficult to know the current best practice for a given area.
> 
>     * It is unclear how to go about the discussion and decision-making process
>       to change a best practice.

Reviewer 4: Overindented, is rendering as a quotation rn
```suggestion
* It is difficult to know the current best practice for a given area.

* It is unclear how to go about the discussion and decision-making process
      to change a best practice.
```

:doc:`OEP-11 <oep-0011-bp-FED-technology>` serves as a great example for the
usefulness of this kind of OEP but limits itself to be relevant only to
frontend technologies.  This OEP would supersede :doc:`OEP-11
<oep-0011-bp-FED-technology>` and encompass not only front-end technology but any
technology that:

>     * Is considered a best practice for the Open edX Platform
> 
>     * Is relevant to a large subset of the Open edX Platform

Reviewer 4: ```suggestion
* Is considered a best practice for the Open edX Platform

* Is relevant to a large subset of the Open edX Platform
```

.. note:: We'll fill this out before the OEP lands but leaving mostly blank for initial feedback.

General Technology Selection
============================

This technology is not specific to frontend or backend code.

#. **Codecov should be used to report code coverage by unit tests**

   **Rationale**: It is important to measure the amount of code covered by our
   automated test suites. By striving for a high level of test coverage, we can
   reduce the number of bugs that can only be found via manual testing, 
   and by using codecov to run coverage in CI,
   contributors are automatically reminded to include tests for any new code.

   The Open edX community has standardized around `Codecov`_ as a coverage reporting service.

#. **Use Github actions for CI Testing**

   **Rationale**: As a healthy open source project we need good continouse integration
   testing to prevent the introduction of obvious bugs into the platform code.  The Github Actions
   tool should be used to perform Continuous Integration(CI) testing for all Open edX repositories
   that perform any sort of software testing.

Frontend Technology Selection
=============================

.. note:: 

>    The phrase “frontend” is used to mean any part of the platform that is shown to

Reviewer 6: Do we actually standardise on "frontend", one word, no punctuation? It's fine if we do, that's just sufficiently unusual that it will probably trigger various spellcheckers and auto-correct tools.

Reviewer 2: As far as I know we do standardize on "frontend." It does seem to trigger spellcheckers, but it also seems to be a very common spelling. For example, [the wiki page for "Frontend and backend"](https://en.wikipedia.org/wiki/Frontend_and_backend) notes:
> the terms frontend and backend (sometimes written as back end or back-end)

   users. This encompasses views rendered in Python on the server, interactive
   interfaces written using JavaScript, and CSS styling.
   
.. _Use React and Redux:

#. **Use React**

>    **Rationale**: React must be used for building new UIs, as it is

Reviewer 6: Is the guidance really MUST on React? Some UIs are going to be so simple that react is overkill. (See above regarding: "new around here"; if this debate has already been had and resolved then I don't need anybody to restart it for me. It's more a question about unexamined assumptions.)

Reviewer 7: @deborahgu If a UI isn't using React, at least one possible concern is the compatibility with using the Paragon design system and React-based UI component library in terms of brand consistency, etc. across UI projects within Open edX.

Reviewer 6: yeah, this is a bigger question than can be resolved in this OEP, but eg. right now we're working on a place where we're dealing with the entire Paragon dependency stack to render 100% static pages, and it's not lightweight. It might be fundamentally worth the cost, but the cost is not small.

>    **Rationale**: React must be used for building new UIs, as it is

Author: that's a fair point, I'm not sure what the intent of the original authors was here, per your second point, MUST feels like a lot here to me as well but I'm not sure what the current stance should be.  This is probably worth a follow-up conversation in the frontend working group.  @brian-smith-tcril @arbrandes @adamstankiewicz what do you think?

>    **Rationale**: React must be used for building new UIs, as it is

Reviewer 8: Yeah, I actually think we're squarely in `SHOULD`-land; even React.  This definitely deserves an dedicated conversation.

   **Exception**: The `edX ESLint Config for ES5`_ may be used only where ES5
   is in use. Both configs may be used in the same codebase through the
   use of `ESLint configuration cascading`_.

#. **JavaScript libraries should be installed via npm**

   **Rationale**: It is important that JavaScript libraries are kept
   up-to-date, and one key component is to make it as simple as possible
   to perform upgrades. Projects that use npm maintain all of their
   library requirements in a single `package.json` file, which can be
   easily changed as the versions change.

#. **JavaScript projects should publish a package-lock file**

   **Rationale**: To keep dependencies up-to-date and builds consistent, JavaScript
   projects should allow patch and minor upgrades in their package.json file and
   commit a package-lock.json file. The package-lock.json file will keep a full list
   of dependencies and their versions, ensuring when built for deployments the
   version of libraries are consistent. This follows the same pattern used in edX
   Python code - see `OEP 18`_ for more information. For more information on
   package-lock files see `Package Lock`_.

> #. **Keep dependencies up to date by using Greenkeeper**

Reviewer 9: Shouldn't this recommend Renovate instead?  See https://openedx.atlassian.net/wiki/spaces/AC/pages/1841791377/Set+up+Renovate+to+Automate+JavaScript+Dependency+Updates#Notes-on-Alternatives for reasons we started to ditch Greenkeeper.  As with the Redux note above we could update it after initial release, but I'd rather not risk people trying to start switching to Greenkeeper after reading this.

Backend Technology Selection
============================

This section will cover any technology or tools that are not considered a part
of the frontend.

.. This section will be empty to begin with and I'll create a separate PR to add any
   technologies that we consider standard today. This will probably include:

    * Django

    * Django Rest Framework

    * Python?

    * drf-yasg (A thing that is currently standard but that I hope to change)

    * tox

    * pytest

    * edx-lint/pylint

Decisions
*********

This section lists the decision records for any time a standard technology has
been change in the system along with the reasons why and the alternatives
considered.

.. We'll have a toctree to an ADR folder with decisions.  We'll move all the existing fronted decisions to under this OEP.

.. toctree::
   :caption: Overall Decisions
   :maxdepth: 1
   :glob:

..   oep-0067/decisions/overall/*

.. toctree::
   :caption: Frontend Decisions
   :maxdepth: 1
   :glob:

..   oep-0067/decisions/frontend/*

.. toctree::
   :caption: Backend Decisions
   :maxdepth: 1
   :glob:

..   oep-0067/decisions/backend/*

Consequences
************

> * OEP-11 will be superseded.

Reviewer 4: Friendly reminder to go into OEP-11 and make this change.
